diff --git a/cpp/ycm/ClangCompleter/TranslationUnit.cpp b/cpp/ycm/ClangCompleter/TranslationUnit.cpp
index 1de1d95..816fc0b 100644
--- a/third_party/ycmd/cpp/ycm/ClangCompleter/TranslationUnit.cpp
+++ b/third_party/ycmd/cpp/ycm/ClangCompleter/TranslationUnit.cpp
@@ -93,11 +93,13 @@ TranslationUnit::TranslationUnit(
                                  ? &cxunsaved_files[ 0 ] : NULL;
 
   // Actually parse the translation unit.
-  CXErrorCode result = clang_parseTranslationUnit2FullArgv(
+  // TODO: Stop stripping argv[0] here and use
+  // clang_parseTranslationUnit2FullArgv, which is available in libclang 3.8.
+  CXErrorCode result = clang_parseTranslationUnit2(
                          clang_index,
                          filename.c_str(),
-                         &pointer_flags[ 0 ],
-                         pointer_flags.size(),
+                         &pointer_flags[ 1 ],
+                         pointer_flags.size() - 1,
                          const_cast<CXUnsavedFile *>( unsaved ),
                          cxunsaved_files.size(),
                          EditingOptions(),
