diff --git a/src/inifile.cpp b/src/inifile.cpp
index d390896..d5044b3 100644
--- a/src/inifile.cpp
+++ b/src/inifile.cpp
@@ -39,7 +39,7 @@ void str2list(const gchar *str, std::list<std::string> &slist)
 {
   gchar *p;
 	slist.clear();
-  while ((p = strchr(str, STRING_SEP))!=NULL) {
+  while ((p = strchr((char*)str, STRING_SEP))!=NULL) {
     slist.push_back(std::string(str, p - str));
     str = p+1;
 	}
diff --git a/src/lib/common.cpp b/src/lib/common.cpp
index e6c2015..fcf5103 100644
--- a/src/lib/common.cpp
+++ b/src/lib/common.cpp
@@ -25,6 +25,8 @@
 #endif
 
 #include "common.hpp"
+#include <string.h>
+#include <stdlib.h>
 
 //looks not optimal, TODO: refactor
 bool DictInfo::load_from_ifo_file(const std::string& ifofilename,
diff --git a/src/lib/data.cpp b/src/lib/data.cpp
index 1fd6994..f05ed44 100644
--- a/src/lib/data.cpp
+++ b/src/lib/data.cpp
@@ -25,6 +25,8 @@
 #include "kmp.h"
 
 #include "data.hpp"
+#include <stdio.h>
+#include <string.h>
 
 
 DictBase::DictBase()
diff --git a/src/lib/treedict.cpp b/src/lib/treedict.cpp
index 9185006..df71f39 100644
--- a/src/lib/treedict.cpp
+++ b/src/lib/treedict.cpp
@@ -28,6 +28,7 @@
 #include "file.hpp"
 
 #include "treedict.hpp"
+#include <string.h>
 
 GtkTreeStore *TreeDict::model=NULL;
 
diff --git a/src/mainwin.cpp b/src/mainwin.cpp
index ab451ad..0abe952 100644
--- a/src/mainwin.cpp
+++ b/src/mainwin.cpp
@@ -35,6 +35,7 @@
 #include "utils.h"
 
 #include "mainwin.h"
+#include <stdlib.h>
 
 
 /**************************************************/
@@ -471,7 +472,7 @@ void TopWin::on_main_menu_about_activate(GtkMenuItem *menuitem, TopWin *oTopWin)
     "Ma Su'an <msa@wri.com.cn>",
         NULL
     };
-    gchar *documenters[] = {
+    const gchar *documenters[] = {
 		"Hu Zheng <huzheng_001@163.com>",
 		"Will Robinson <wsr23@stanford.edu>",
 		"Anthony Fok <foka@debian.org>",
@@ -1420,13 +1421,14 @@ gboolean TextWin::OnSearchKeyPress(GtkWidget *widget, GdkEventKey *event,
 void TextWin::OnFindSearchPanel(GtkWidget *widget, TextWin *oTextWin)
 {
   const gchar *text=gtk_entry_get_text(oTextWin->eSearch);
-  if (text && *text)
+  if (text && *text) {
     if (oTextWin->Find(text, oTextWin->search_from_beginning)) {
       oTextWin->search_from_beginning = FALSE;
 
       oTextWin->find_text = text;
     } else
       oTextWin->search_from_beginning = TRUE;
+  }
   
 }
 
diff --git a/src/pangoview.cpp b/src/pangoview.cpp
index ed849d9..3ac6690 100644
--- a/src/pangoview.cpp
+++ b/src/pangoview.cpp
@@ -26,6 +26,7 @@
 #include "utils.h"
 
 #include "pangoview.h"
+#include <string.h>
 
 void PangoView::Create(bool autoresize_)
 {
diff --git a/src/prefsdlg.cpp b/src/prefsdlg.cpp
index 8a04471..00eb6ac 100644
--- a/src/prefsdlg.cpp
+++ b/src/prefsdlg.cpp
@@ -58,7 +58,7 @@ enum
 
 
 struct CategoriesTreeItem {
-  gchar			*category;
+  const gchar			*category;
 	
   CategoriesTreeItem 	*children;
   
@@ -1158,7 +1158,7 @@ void PrefsDlg::on_setup_mainwin_searchwebsite_add_button_clicked(GtkWidget *widg
 															 GTK_WINDOW(searchwebsite_add_dialog),
 																 (GtkDialogFlags) (GTK_DIALOG_MODAL | GTK_DIALOG_DESTROY_WITH_PARENT),
 															 GTK_MESSAGE_INFO,	GTK_BUTTONS_OK,
-															 error_msg);
+															 "%s", error_msg);
 			
 			gtk_dialog_set_default_response(GTK_DIALOG(message_dlg), GTK_RESPONSE_OK);
 			gtk_window_set_resizable(GTK_WINDOW(message_dlg), FALSE);
diff --git a/src/x11_iskeyspressed.hpp b/src/x11_iskeyspressed.hpp
index 9b0d375..082fee5 100644
--- a/src/x11_iskeyspressed.hpp
+++ b/src/x11_iskeyspressed.hpp
@@ -4,6 +4,7 @@
 #include <gdk/gdkx.h>
 #include <X11/keysym.h>
 #include <gtk/gtk.h>
+#include <memory>
 
 #include "iskeyspressed.hpp"
 
diff --git a/tests/t_config_file.cpp b/tests/t_config_file.cpp
index f136cac..f5ee9ff 100644
--- a/tests/t_config_file.cpp
+++ b/tests/t_config_file.cpp
@@ -29,6 +29,7 @@
 #include <iterator>
 #include <iostream>
 #include <gtk/gtk.h>
+#include <memory>
 
 #include "config_file.hpp"
 #include "inifile.hpp"
