diff --git a/js/xpconnect/src/XPCWrappedNative.cpp b/js/xpconnect/src/XPCWrappedNative.cpp
index a12e36ba..59f85996 100644
--- a/js/xpconnect/src/XPCWrappedNative.cpp
+++ b/js/xpconnect/src/XPCWrappedNative.cpp
@@ -2218,8 +2218,9 @@ XPCWrappedNative::ToString(XPCWrappedNativeTearOff* to /* = nullptr */ ) const
                 const char* fmt = (i == 0) ?
                                     "(%s" : (i == count-1) ?
                                         ", %s)" : ", %s";
+                const char * arg = array[i]->GetNameString();
                 name = JS_sprintf_append(name, fmt,
-                                         array[i]->GetNameString());
+                                         arg ? arg : "(null)");
             }
         }
     }
